### PR TITLE
Add lock and new tests

### PR DESCRIPTION
**Pull request by GabrielLyonB**
## Objective
- Add `lock!`s to the methods that involve change in the nodes queue. I included `self.lock!` because if we use it before each transaction involving the update/create/delete of some of the nodes, we'll be preventing multiple process accessing and updating the nodes. 
- I change the `get_next_in_queue` to `get_next_node_in_queue` to return the next node and change the `get_next_in_queue` to return the next nodable. Notice: that both methods updates the queue.
- Add tests for the `formatted_queue` method.

### CODE REVIEW

**Review by ldlsegovia — APPROVED** [review]
👏🏻

File: app/models/concerns/queue_it/queable.rb [hunk 1]
```diff
@@ -31,32 +35,32 @@ def get_next_in_queue
       end
     end
 
-    def formatted_queue(node_attribute)
+    def formatted_queue(nodable_attribute)
```
Comment (monkeyci) [inline app/models/concerns/queue_it/queable.rb:38]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Metrics/AbcSize: Assignment Branch Condition size for formatted_queue is too high. [<3, 26, 5> 26.65/25]

File: app/models/concerns/queue_it/queable.rb [hunk 2]
```diff
@@ -31,32 +35,32 @@ def get_next_in_queue
...
       return if local_queue.empty?
 
       if local_queue.one_node?
-        [local_queue.head_node.nodable.send(node_attribute)]
+        [local_queue.head_node.nodable.send(nodable_attribute)]
       elsif local_queue.two_nodes?
-        [local_queue.head_node.nodable.send(node_attribute) +
-          local_queue.tail_node.nodable.send(node_attribute)]
+        [local_queue.head_node.nodable.send(nodable_attribute),
+          local_queue.tail_node.nodable.send(nodable_attribute)]
```
Comment (monkeyci) [inline app/models/concerns/queue_it/queable.rb:45]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Layout/ArrayAlignment: Align the elements of an array literal if they span more than one line.

File: app/models/queue_it/queue.rb [hunk 3]
```diff
@@ -29,8 +29,9 @@ def two_nodes?
 
     def get_next_in_queue_with_length_two
       next_node = ActiveRecord::Base.transaction do
-        old_head_node = head_node
-        old_tail_node = tail_node
+        self.lock!
```
Comment (monkeyci) [inline app/models/queue_it/queue.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Style/RedundantSelf: Redundant self detected.

File: app/models/queue_it/queue.rb [hunk 4]
```diff
@@ -41,9 +42,10 @@ def get_next_in_queue_with_length_two
 
     def get_next_in_queue_generic
       next_node = ActiveRecord::Base.transaction do
-        old_head_node = head_node
-        old_second_node = old_head_node.child_node
-        old_tail_node = tail_node
+        self.lock!
```
Comment (monkeyci) [inline app/models/queue_it/queue.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Style/RedundantSelf: Redundant self detected.

File: app/models/queue_it/queue.rb [hunk 5]
```diff
@@ -52,17 +54,28 @@ def get_next_in_queue_generic
       next_node
     end
 
-    def push_node_when_queue_lenght_is_one(nodable, in_head)
+    def push_node_when_queue_length_is_zero(nodable)
+      ActiveRecord::Base.transaction do
+        self.lock!
```
Comment (monkeyci) [inline app/models/queue_it/queue.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Style/RedundantSelf: Redundant self detected.

File: app/models/queue_it/queue.rb [hunk 6]
```diff
@@ -52,17 +54,28 @@ def get_next_in_queue_generic
...
+      end
+    end
+
+    def push_node_when_queue_length_is_one(nodable, in_head)
       if in_head
         push_in_head(nodable)
       else
-        nodes.create!(nodable: nodable, kind: :tail, parent_node: head_node)
+        ActiveRecord::Base.transaction do
+          self.lock!
```
Comment (monkeyci) [inline app/models/queue_it/queue.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Style/RedundantSelf: Redundant self detected.

File: app/models/queue_it/queue.rb [hunk 7]
```diff
@@ -52,17 +54,28 @@ def get_next_in_queue_generic
...
+          nodes.create!(nodable: nodable, kind: :tail, parent_node: head_node)
+        end
       end
     end
 
     def push_in_head(nodable)
       ActiveRecord::Base.transaction do
-        old_head_node = head_node
+        self.lock!
```
Comment (monkeyci) [inline app/models/queue_it/queue.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Style/RedundantSelf: Redundant self detected.

File: app/models/queue_it/queue.rb [hunk 8]
```diff
@@ -72,7 +85,8 @@ def push_in_head(nodable)
 
     def push_in_tail(nodable)
       ActiveRecord::Base.transaction do
-        old_tail_node = tail_node
+        self.lock!
```
Comment (monkeyci) [inline app/models/queue_it/queue.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Style/RedundantSelf: Redundant self detected.

File: spec/concerns/queue_it/queable_spec.rb [hunk 9]
```diff
@@ -138,23 +138,78 @@
...
-      it "" do
-        expect(task.get_next_in_queue.nodable).to eq(third_nodable)
+      it "expects to get third nodable and to correctly move the queue" do
+        expect(task.get_next_node_in_queue.nodable).to eq(third_nodable)
         expect(task.queue.nodes.find_by(kind: :head).nodable).to eq(second_nodable)
         expect(task.queue.nodes.find_by(kind: :any).nodable).to eq(first_nodable)
         expect(task.queue.nodes.find_by(kind: :tail).nodable).to eq(third_nodable)
       end
     end
+
```
Comment (monkeyci) [inline spec/concerns/queue_it/queable_spec.rb:148]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body end.

File: spec/concerns/queue_it/queable_spec.rb [hunk 10]
```diff
@@ -138,23 +138,78 @@
...
   end
 
-  describe '#empty_queue' do
-    describe 'when queue is empty' do
+  describe '#delete_queue_nodes' do
+    context 'when queue is empty' do
       before { task.find_or_create_queue! }
 
-      it { expect { task.empty_queue }.not_to change { task.queue.size } }
+      it { expect { task.delete_queue_nodes }.not_to change { task.queue.size } }
```
Comment (monkeyci) [inline spec/concerns/queue_it/queable_spec.rb:209]: ⚠️ **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Lint/AmbiguousBlockAssociation: Parenthesize the param change { task.queue.size } to make sure that the block will be associated with the change method call.

**Review by ldlsegovia — APPROVED** [review]

**Pull request merged**